### DISCUSSION TITLE
Streaming 24-bit audio

### DISCUSSION THREAD

**Discussion by geoffw64** [snapcast/snapcast#1256] [Q&A]
Hi all, 

I'm enjoying Snapcast and Mopidy and need a bit of help streaming higher resolution audio. 

Environment: 
Snapcast 0.28.0, Snapweb 0.7.0, Mopidy 3.4.1 and Iris 3.69.3, 
Installed on a single RPi 4B running RaspiOS Bookworm. 
It all works great when I configure 44100:16:2 or 96000:16:2.  
I can also stream a 96kHz 24-bit stereo flac file using a 48000:24:2 configuration

However, when I try to use 24 bit resolution Snap will not play the music. 
A dialog box containing the following text pops up in Snapcast web:
"Stream must be stereo with 16 bit depth, actual format: 48000:32:2"
I get the same behaviour on Safari (MAC OS and iOS) and Chromium on RaspiOS.  
I haved tried 24 bit configuration and the result is the same. 

My snapserver.conf contains:
[server]
[http]
doc_root = /usr/share/snapweb
host=192.168.1.34
[tcp]
[stream]
source = pipe:///run/snapserver/snapfifo?name=default&mode=read&codec=pcm&sampleformat=48000:24:2&chunk_ms=80&buffer=2000
[streaming_client]
initial_volume = 50
[logging]

My mopidfy.conf file contains: 
[http]
hostname = 0.0.0.0 
[audio] 
mixer_volume = 50 
output = audioconvert ! audioresample ! audio/x-raw,rate=48000,channels=2,format=S24_32LE ! filesink location=/run/snapserver/snapfifo
[file]
media_dirs = /media/geoff/MEDIA/Music
[local]
enabled = true ; 
media_dir = /media/geoff/MEDIA/Music 
[logging]
verbosity = 4

My effective Mopiody config is: 

INFO     [MainThread] mopidy.__main__ Starting Mopidy 3.4.1
DEBUG    [MainThread] mopidy.ext Loading entry point: iris = mopidy_iris:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-Iris 3.69.3
DEBUG    [MainThread] mopidy.ext Loading entry point: local = mopidy_local:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-Local 3.2.1
DEBUG    [MainThread] mopidy.ext Loading entry point: file = mopidy.file:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-File 3.4.1
DEBUG    [MainThread] mopidy.ext Loading entry point: http = mopidy.http:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-HTTP 3.4.1
DEBUG    [MainThread] mopidy.ext Loading entry point: m3u = mopidy.m3u:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-M3U 3.4.1
DEBUG    [MainThread] mopidy.ext Loading entry point: softwaremixer = mopidy.softwaremixer:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-SoftwareMixer 3.4.1
DEBUG    [MainThread] mopidy.ext Loading entry point: stream = mopidy.stream:Extension
DEBUG    [MainThread] mopidy.ext Loaded extension: Mopidy-Stream 3.4.1
DEBUG    [MainThread] mopidy.ext Discovered extensions: iris, local, file, http, m3u, softwaremixer, stream
DEBUG    [MainThread] mopidy.config.keyring Fetching passwords from your keyring failed. Any passwords stored in the keyring will not be available. (org.freedesktop.DBus.Error.Spawn.ExecFailed: /usr/bin/dbus-launch terminated abnormally with the following error: Authorization required, but no authorization protocol specified

Autolaunch error: X11 initialization failed.
)
INFO     [MainThread] mopidy.config Loading config from builtin defaults
INFO     [MainThread] mopidy.config Loading config from file:///usr/share/mopidy/conf.d/mopidy.conf
INFO     [MainThread] mopidy.config Loading config from file:///etc/mopidy/mopidy.conf
INFO     [MainThread] mopidy.config Loading config from command line options
DEBUG    [MainThread] mopidy.ext Validating extension: iris
DEBUG    [MainThread] mopidy.ext Validating extension: local
DEBUG    [MainThread] mopidy.ext Validating extension: file
DEBUG    [MainThread] mopidy.ext Validating extension: http
DEBUG    [MainThread] mopidy.ext Validating extension: m3u
DEBUG    [MainThread] mopidy.ext Validating extension: softwaremixer
DEBUG    [MainThread] mopidy.ext Validating extension: stream
INFO     [MainThread] mopidy.__main__ Enabled extensions: stream, file, softwaremixer, http, m3u, local, iris
INFO     [MainThread] mopidy.__main__ Disabled extensions: none
[core]
cache_dir = /var/cache/mopidy
config_dir = /etc/mopidy
data_dir = /var/lib/mopidy
max_tracklist_length = 10000
restore_state = false

[logging]
verbosity = 4
format = %(levelname)-8s [%(threadName)s] %(name)s %(message)s
color = false
config_file = 

[audio]
mixer = software
mixer_volume = 50
output = audioconvert ! audioresample ! audio/x-raw,rate=48000,channels=2,format=S24_32LE ! filesink location=/run/snapserver/snapfifo
buffer_time = 

[proxy]
scheme = 
hostname = 
port = 
username = 
password = 

[iris]
enabled = true
country = NZ
locale = en_NZ
verify_certificates = true
snapcast_enabled = true
snapcast_host = localhost
snapcast_port = 1780
snapcast_ssl = false
snapcast_stream = Default
spotify_authorization_url = https://jamesbarnsley.co.nz/iris/auth_spotify.php
lastfm_authorization_url = https://jamesbarnsley.co.nz/iris/auth_lastfm.php
genius_authorization_url = https://jamesbarnsley.co.nz/iris/auth_genius.php
data_dir = $XDG_DATA_DIR/iris

[local]
enabled = true
max_search_results = 100
media_dir = /media/geoff/MEDIA/Music
scan_timeout = 1000
scan_flush_threshold = 100
scan_follow_symlinks = false
included_file_extensions = 
excluded_file_extensions = 
  .cue
  .directory
  .html
  .jpeg
  .jpg
  .log
  .nfo
  .pdf
  .png
  .txt
  .zip
directories = 
  Albums                  local:directory?type=album
  Artists                 local:directory?type=artist
  Composers               local:directory?type=artist&role=composer
  Genres                  local:directory?type=genre
  Performers              local:directory?type=artist&role=performer
  Release Years           local:directory?type=date&format=%25Y
  Tracks                  local:directory?type=track
  Last Week's Updates     local:directory?max-age=604800
  Last Month's Updates    local:directory?max-age=2592000
timeout = 10
use_artist_sortname = false
album_art_files = 
  *.jpg
  *.jpeg
  *.png

[file]
enabled = true
media_dirs = 
  /media/geoff/MEDIA/Music
excluded_file_extensions = 
  .directory
  .html
  .jpeg
  .jpg
  .log
  .nfo
  .pdf
  .png
  .txt
  .zip
show_dotfiles = false
follow_symlinks = false
metadata_timeout = 1000

[http]
enabled = true
hostname = 0.0.0.0
port = 6680
zeroconf = Mopidy HTTP server on $hostname
allowed_origins = 
csrf_protection = true
default_app = mopidy

[m3u]
enabled = true
base_dir = 
default_encoding = latin-1
default_extension = .m3u8
playlists_dir = 

[softwaremixer]
enabled = true

[stream]
enabled = true
protocols = 
  http
  https
  mms
  rtmp
  rtmps
  rtsp
metadata_blacklist = 
timeout = 5000

I can see that mopidy is playing the audio in the system journal. 

Thx in advance for any help. 

Cheers, Geoff.

**Comment by SantiagoSotoC** [comment]
Snapweb does not support more than 16 bit audio

**Reply by badaix** [reply]
It actually does support hi-res audio since last Sunday, introduced with [this commit](https://github.com/badaix/snapweb/commit/75c2674d62c96cd9ad4228fc418291b75693ed19). An automated build of version `v0.7.50` is available here: https://github.com/badaix/snapweb/actions/runs/9748384049, see also issue #1248

**Reply by geoffw64** [reply]
Brilliant!  I will give it a test.  
Thanks so much for this great software.  😀

**Comment by geoffw64** [comment]
Did a bit more investigation...  I can stream 96kHz 32-bit audio to snapclient, so my setup is working. 
Perhaps snapweb can't stream 24-bit (or higher) resolution. 

Has anyone else tried using snapweb for high resolution audio?

**Reply by badaix** [reply]
Flac will limit you to 24 bit, Snapweb can play hi-res audio starting with the current development version v0.7.50, see my [comment above](https://github.com/badaix/snapcast/discussions/1256#discussioncomment-9933640)

**Reply by geoffw64** [reply]
Wondering why there would be a 24-bit limit with flac?  
The flac protocol itself supports 32.
Cheers,
Geoff.

**Reply by badaix** [reply]
Looks like 32 bit support has been added to Flac 1.4 in 2022. The Linux Mint 21.3 on my dev machine has an older version installed, my RPi has the 1.4 installed, so it could actually work there, since it's not limited anywhere in the source code, flac will just return an error, causing Snapserver to terminate itself: 

```
2024-07-02 11-09-02.555 [Error] (Snapserver) Exception: ERROR: initializing encoder: FLAC__STREAM_ENCODER_INIT_STATUS_INVALID_BITS_PER_SAMPLE
```

**Reply by badaix** [reply]
I just checked: Snapweb is using [libflacjs](https://www.npmjs.com/package/libflacjs), which is based on libflac 1.3.2, i.e does not support 32 bit.

**Reply by geoffw64** [reply]
No worries, thanks for letting me know.  
The highest resolution content I have is 24 bit in any case.   I tried 32 because 24 didnt work, probably due to an unrelated config error on my part.    
Mopidy, Snapserver and Snapcast seem happy with 32-bit, although I’ve only used 16- and 24- bit source material.  
Cheers, Geoff.
